contracts/helper-data/HelperDataFactory.sol

```diff
@@ -0,0 +1,353 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.28;
+
+import {ERC1967Utils} from "@openzeppelin/contracts/proxy/ERC1967/ERC1967Utils.sol";
+import {EnumerableSet} from "@openzeppelin/contracts/utils/structs/EnumerableSet.sol";
+import {EnumerableMap} from "@openzeppelin/contracts/utils/structs/EnumerableMap.sol";
+
+import {OwnableUpgradeable} from "@openzeppelin/contracts-upgradeable/access/OwnableUpgradeable.sol";
+import {UUPSUpgradeable} from "@openzeppelin/contracts-upgradeable/proxy/utils/UUPSUpgradeable.sol";
+
+import {SSTORE2} from "solady/src/utils/SSTORE2.sol";
+
+import {ADeployerGuard} from "@solarity/solidity-lib/utils/ADeployerGuard.sol";
+import {Paginator} from "@solarity/solidity-lib/libs/arrays/Paginator.sol";
+import {Vector} from "@solarity/solidity-lib/libs/data-structures/memory/Vector.sol";
+
+import {IHelperDataFactory} from "../interfaces/helper-data/IHelperDataFactory.sol";
+
+contract HelperDataFactory is
+    IHelperDataFactory,
+    OwnableUpgradeable,
+    UUPSUpgradeable,
+    ADeployerGuard
+{
+    using Paginator for *;
+    using EnumerableSet for *;
+    using EnumerableMap for EnumerableMap.UintToAddressMap;
+    using Vector for Vector.AddressVector;
+
+    bytes32 private constant HELPER_DATA_FACTORY_STORAGE_SLOT =
+        keccak256("unforgettable.contract.helper.data.helper.data.factory.storage");
+
+    struct HelperDataFactoryStorage {
+        EnumerableSet.AddressSet helperDataManagers;
+        EnumerableSet.AddressSet registeredAccounts;
+        mapping(address => AccountData) accountsData;
+    }
+
+    modifier onlyHelperDataManager() {
+        _checkHelperDataManager(msg.sender);
+        _;
+    }
+
+    modifier onlyRegisteredAccount(address account_) {
+        _onlyRegisteredAccount(account_);
+        _;
+    }
+
+    constructor() ADeployerGuard(msg.sender) {
+        _disableInitializers();
+    }
+
+    function _getHelperDataFactoryStorage()
+        private
+        pure
+        returns (HelperDataFactoryStorage storage _hdfs)
+    {
+        bytes32 slot_ = HELPER_DATA_FACTORY_STORAGE_SLOT;
+
+        assembly ("memory-safe") {
+            _hdfs.slot := slot_
+        }
+    }
+
+    function initialize(address[] calldata initialManagers_) external initializer onlyDeployer {
+        __Ownable_init(msg.sender);
+
+        _addHelperDataManagers(initialManagers_);
+    }
+
+    /// @inheritdoc IHelperDataFactory
+    function addHelperDataManagers(address[] calldata managersToAdd_) external onlyOwner {
+        _addHelperDataManagers(managersToAdd_);
+    }
+
+    /// @inheritdoc IHelperDataFactory
+    function removeHelperDataManagers(address[] calldata managersToRemove_) external onlyOwner {
+        _removeHelperDataManagers(managersToRemove_);
+    }
+
+    /// @inheritdoc IHelperDataFactory
+    function registerAccount(
+        address account_,
+        uint256 partIndex_,
```

Comment:
So we don't check that the partIndex already exists and that the ordering is sequential?